### PR TITLE
feat: (IAC-355) Enable Multi-tenant deployments

### PR DESCRIPTION
**Pull request by riragh**
# Changes
Enabling DAC to do multi-tenant deployments. Tasks were divided into different tickets. 
- IAC-440: Initial code setup to support MT
- IAC-441: README updates for Multi-tenancy support
- IAC-454: Onboard - Offboard Tenant
- IAC-492: OpenLDAP setup required for MT
- IAC-597: Increase max_connections for internal Crunchy PG
- IAC-601: Enhancement: Tenant CAS customization

# Tests
See the above mentioned internal tickets for testing

### CODE REVIEW

**Comment by riragh** [comment]
Variable names updated: SAS_TENANT_IDS is being renamed to V4MT_TENANT_IDS and SAS_PROVIDER_PASSWORD is renamed to V4MT_PROVIDER_PASSWORD.
Files were updated to be all lower case and use hyphens instead of underscore.
README documents now SAS Viya Multi-tenancy instead of only Multi-tenancy.

**Review by thpang — COMMENTED** [review]
The more I think about this one, the actions should have been name spaced as we have with other tasks:

baseline - performs baseline items
viya - performs viya items

There should be a multi-tenancy item:

mt - performs multi-tenancy items. This would have actions like: onboard, cas-onboard, etc

At this point we'd have to retest but again, long term thinking and pattern matching with the code base would have made this seem like the logical choice. As it is now we just run `onboard` or `cas-onboard` they have no real meaning and or context.

If you look at the doc [here](https://github.com/sassoftware/viya4-deployment/blob/main/docs/user/AnsibleUsage.md#running). You can see there are `Actions` and `Tasks` with this new implementation things are broken. There are no more actions or tasks.

**Comment by riragh** [comment]
> The more I think about this one, the actions should have been name spaced as we have with other tasks:
> 
> baseline - performs baseline items viya - performs viya items
> 
> There should be a multi-tenancy item:
> 
> mt - performs multi-tenancy items. This would have actions like: onboard, cas-onboard, etc
> 
> At this point we'd have to retest but again, long term thinking and pattern matching with the code base would have made this seem like the logical choice. As it is now we just run `onboard` or `cas-onboard` they have no real meaning and or context.
> 
> If you look at the doc [here](https://github.com/sassoftware/viya4-deployment/blob/main/docs/user/AnsibleUsage.md#running). You can see there are `Actions` and `Tasks` with this new implementation things are broken. There are no more actions or tasks.

In the new change, updated the actions and tasks. Task is `multi-tenancy` and actions are `onboard`, `cas-onboard` and `offboard`. Ran different scenarios to cover this new change.

File: examples/multi-tenancy/openldap/openldap-modify-mt-users-groups.yaml [hunk 1]
```diff
@@ -0,0 +1,207 @@
+---
```
[the remark this answers is not in the document]
Comment (riragh) [reply examples/multi-tenancy/openldap/openldap-modify-mt-users-groups.yaml:14]: This file only contains the provider. Added the diagram for the same. To add tenants you need to follow the steps in the Multi-Tenancy Readme from section `OpenLDAP Customizations`.

File: docs/user/Multi-Tenancy.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# Multi-Tenancy
```
[the remark this answers is not in the document]
Comment (riragh) [reply docs/user/Multi-Tenancy.md]: This was updated in the last change.

File: docs/CONFIG-VARS.md [hunk 3]
```diff
@@ -333,3 +334,15 @@ The nfs-client is currently supported by the newer nfs-subdir-external-provision
...
+| SAS_TENANT_IDS | One or more tenant IDs to onboard or offboard | string | | false | Example: Single tenant ID: "acme" or Multiple tenant IDs: "acme, cyberdyne, intech" | onboard, cas-onboard, offboard |
+| SAS_PROVIDER_PASSWORD | Optional: The password that is applied to the tenant administrator on each onboarded tenant | string | | false | When SAS_PROVIDER_PASSWORD is specified SAS_PROVIDER_PASSWORD_{{TENANT-ID}} can not be used. See details [here](https://go.documentation.sas.com/doc/en/itopscdc/v_029/caltenants/p0emzq13c0zbhxn1hktsdlmig934.htm#p1ghvmezrb3cvxn1h7vg4uguqct6) | onboard, cas-onboard, offboard |
+| SAS_PROVIDER_PASSWORD_{{TENANT-ID}} | Optional: Unique sasprovider password for each tenant being onboarded| string | | false | When SAS_PROVIDER_PASSWORD_{{TENANT-ID}} is specified SAS_PROVIDER_PASSWORD can not be used. See details [here](https://go.documentation.sas.com/doc/en/itopscdc/v_029/caltenants/p0emzq13c0zbhxn1hktsdlmig934.htm#p1ghvmezrb3cvxn1h7vg4uguqct6) | onboard, cas-onboard, offboard |
```
[the remark this answers is not in the document]
Comment (riragh) [reply docs/CONFIG-VARS.md]: The variable names are updated for conformity.

**Review by thpang — APPROVED** [review]
Thanks so much for the update. Adding `multi-tenancy` is much better. LGTM

**Review by jarpat — APPROVED** [review]

**Review by sayeun — COMMENTED** [review]
Talked with Ritika, there are some doc links need to be updated.

**Comment by riragh** [comment]
> Talked with Ritika, there are some doc links need to be updated.

Done. Updated all links to always point to default cadence.

**Review by sayeun — APPROVED** [review]
Looks 👍

**Pull request merged**